File: Javascript/ShortestPathInBinaryMatrix.js
var shortestPathBinaryMatrix = function(grid){
    const [ROWS,COLS] = [grid.length, grid[0].length]
    const queue = []
    const visit = new Set()
    let length = 0
    queue.push([0,0])
    visit.add(`00`)

    if(grid[0][0] == 1){
        return -1
    }

    while(queue.length>0){
        const queueLength = queue.length
        for(let i=0;i<queueLength;i++){
            const [r,c] = queue.shift()

            if (r === ROWS-1 && c === COLS -1){
                length+=1
                return length
            }

            const directions = [[0,1],[0,-1],[1,0],[-1,0],[1,1],[1,-1],[-1,1],[-1,-1]]
            for (const [dr,dc] of directions){
                if (r+dr<0 || c+dc<0 || r+dr>=ROWS || c+dc>=COLS || 
                    grid[r+dr][c+dc]===1 || visit.has(`${r+dr}${c+dc}`)){
                    continue
                }
                queue.push([r+dr,c+dc])
                visit.add(`${r+dr}${c+dc}`)
            }   
        }
        length+=1 
        
    }

    return -1
}

File: Python/ShortestPathInBinaryMatrix.py
from collections import deque

def shortestPathBinaryMatrix(grid):
    ROWS,COLS = len(grid),len(grid[0])
    queue = deque()
    visit = set()
    length = 0
    queue.append((0,0))
    visit.add((0,0))

    if grid[0][0] == 1:
        return -1

    while queue:
        for i in range(len(queue)):
            r,c = queue.popleft()

            if r == ROWS-1 and c == COLS-1:
                length+=1
                return length
            
            directions = [[1,0],[-1,0],[0,1],[0,-1],[1,1],[1,-1],[-1,1],[-1,-1]]
            for dr,dc in directions:
                if r+dr<0 or c+dc<0 or r+dr>=ROWS or c+dc>=COLS or \
                    grid[r+dr][c+dc] == 1 or (r+dr,c+dc) in visit:
                    continue

                queue.append((r+dr,c+dc))
                visit.add((r+dr,c+dc))

        length +=1

    return -1